OrbitCaptureClient/CaptureEventProcessor.cpp

```diff
@@ -340,3 +428,254 @@ void CaptureEventProcessor::SendTracepointInfoToListenerIfNecessary(
+      } else {
+        matching_begin_submission = FindMatchingGpuQueueSubmission(
+            begin_marker_thread_id, begin_marker_post_submission_cpu_timestamp);
+      }
+
+      // Note, we receive submissions of a single queue in order (by CPU submission time). So if
+      // there is no matching "begin submission", the "begin" was submitted before the "end" and
+      // we lost the record of the "begin's" submission (which should not happen).
+      CHECK(matching_begin_submission != nullptr);
+
+      std::optional<GpuCommandBuffer> begin_submission_first_command_buffer =
+          FindFirstCommandBuffer(*matching_begin_submission);
+      CHECK(begin_submission_first_command_buffer.has_value());
+
+      const GpuJob* matching_begin_job = FindMatchingGpuJob(
+          begin_marker_thread_id, begin_marker_meta_info.pre_submission_cpu_timestamp(),
+          begin_marker_post_submission_cpu_timestamp);
+      CHECK(matching_begin_job != nullptr);
+
+      marker_timer.set_start(completed_marker.begin_marker().gpu_timestamp_ns() +
+                             matching_begin_job->gpu_hardware_start_time_ns() -
+                             begin_submission_first_command_buffer->begin_gpu_timestamp_ns());
+      if (begin_marker_thread_id == gpu_queue_submission.meta_info().tid()) {
+        marker_timer.set_thread_id(begin_marker_thread_id);
+      }
+
+      DecrementUnprocessedBeginMarkers(begin_marker_thread_id,
+                                       begin_marker_post_submission_cpu_timestamp);
+    } else {
+      marker_timer.set_start(begin_capture_time_ns_);
+      marker_timer.set_thread_id(-1);
+    }
+
+    marker_timer.set_depth(completed_marker.depth());
+    marker_timer.set_timeline_hash(timeline_marker_hash);
+    marker_timer.set_processor(-1);
+    marker_timer.set_type(TimerInfo::kGpuDebugMarker);
+    marker_timer.set_end(completed_marker.end_gpu_timestamp_ns() -
+                         first_command_buffer->begin_gpu_timestamp_ns() +
+                         matching_gpu_job.gpu_hardware_start_time_ns());
+
+    CHECK(string_intern_pool_.contains(completed_marker.text_key()));
+    const std::string& text = string_intern_pool_.at(completed_marker.text_key());
+    uint64_t text_key = GetStringHashAndSendToListenerIfNecessary(text);
+
+    if (completed_marker.has_color()) {
+      Color* color = marker_timer.mutable_color();
+      color->set_red(static_cast<uint32_t>(completed_marker.color().red() * 255));
+      color->set_green(static_cast<uint32_t>(completed_marker.color().green() * 255));
+      color->set_blue(static_cast<uint32_t>(completed_marker.color().blue() * 255));
+      color->set_alpha(static_cast<uint32_t>(completed_marker.color().alpha() * 255));
+    }
+    marker_timer.set_user_data_key(text_key);
+    capture_listener_->OnTimer(marker_timer);
+  }
+}
+
+std::optional<orbit_grpc_protos::GpuCommandBuffer> CaptureEventProcessor::FindFirstCommandBuffer(
```

Comment:
I would say this is "get" more than "find". Maybe "extract" if you really don't want to use "get".